### ISSUE TITLE
Obligatory use of the existing definitions

### ISSUE THREAD

**Issue by vladimirias** [UniMath/UniMath#395]
I suggest the following addition to our rules:

1. Formalizing a result or a group of results one must use the existing definitions wherever they are available. For example, if the ring of integers appears in the formulation of a theorem or in a definition of a more complex object the existing the definition of [hz] must be used.

If a different model of [hz] is more convenient for the proof, including intermediate lemmas, such a definition may be introduced and used as long as a proof of the theorem formulated using the existing definition is also produced.

**Comment by DanGrayson** [comment]
In general, this sounds like a good principle, but I have some code for limits in categories based on representable functors in `Ktheory` that duplicates some code in `CategoryTheory` (with a different approach), and I'd like to keep it.

**Comment by vladimirias** [comment]
Which particular definitions are duplicated?

**Comment by DanGrayson** [comment]
at least initial object, terminal object, zero object, coproduct, product, direct sum, kernel, cokernel, binary products, binary coproducts, limits, colimits, cone, pushout, pullback.

**Comment by DanGrayson** [comment]
My main motivation was to reduce duplication of proofs.  In this case, uniqueness of universal constructions up to isomorphism is proved just once, as uniqueness of the object representing a functor.

**Comment by DanGrayson** [comment]
I'd be happy to label my development as experimental somehow, so others don't use it.  Meanwhile, if @tpannila has the energy to formalize exact categories (in the sense of Quillen) after he does abelian categories, that would save me a bunch of trouble, and I could base my K-theory on that.

**Comment by tpannila** [comment]
@DanGrayson , I have some problems in formalizing abelian categories. If
these problems are difficult to solve, as I expect, then it takes a lot of time to
formalize abelian categories. I don't know anything about exact categories,
but I can take a look at them after we have formalized abelian categories.

**Comment by benediktahrens** [comment]
On 07/03/2016 02:07 PM, Daniel R. Grayson wrote:
> My main motivation was to reduce duplication of proofs.  In this
> case, uniqueness of universal constructions up to isomorphism is
> proved just once, as uniqueness of the object representing a
> functor.

It is one of the fundamental principles of category theory that every
concept can be expressed in multiple different ways.
The different concepts might be useful in different situations.

Also, it might turn out that some concepts work under fewer assumptions
than others. For instance, the homs-are-sets assumption is not necessary
for direct formalizations of specific limits, but necessary for working
with (co)cones.
Hence, things that are equivalent classically can turn out to be not
equivalent as types in univalent foundations.

Let us also not forget that equivalence of types is rather coarse.
(E.g., the interval is contractible, but from an interval one obtains
function extensionality.
Another example is given by different representations of lists, some of
which have a normal form, whereas others haven't.)
So there are situations where one wants to work with one specific type
out of many equivalent ones.


Certainly, usage of the existing definitions is encouraged. That's in
everybody's interest, since it allows usage of existing theorems.
If, however, someone thinks that the exisiting stuff doesn't fit their
need, they should be free to add an equivalent definition.


In short: I don't think that there should be a rule about this. Having
more rules makes less people interested in contributing.
We should have as many rules as necessary, but as few as possible.

**Comment by vladimirias** [comment]
Then I suggest the following modified version of the rule:

"Formalizing a result or a group of results one should use the existing definitions wherever they are available. For example, if the ring of integers appears in the formulation of a theorem or in a definition of a more complex object the existing the definition of [hz] must be used.

If, for some reason, a different model of a construction is strongly preferable it must be proved to be equivalent to the existing one."

**Comment by DanGrayson** [comment]
Sounds okay to me.

Will people be dismayed to learn that this definition exists in Ktheory?:
```
Definition Precategory := Σ C:precategory, has_homsets C.
```

**Comment by vladimirias** [comment]
No, but it is not a duplication since we don’t have this concept anywhere else. Right?

I am a little concerned with the use of the capital letter as a device indicating the addition of another property. It is a good device in general but if we use it it would be good to have some consistency in how we use it.

**Comment by DanGrayson** [comment]
I once made a proposal for what capitalization should mean: https://github.com/UniMath/UniMath/issues/176

**Comment by DanGrayson** [comment]
Re: "No, but it is not a duplication since we don’t have this concept anywhere else. Right?"

Well, we might have to move it upstream.  I'd be happy to rename it.

**Comment by benediktahrens** [comment]
On 07/04/2016 05:21 PM, Vladimir Voevodsky wrote:
> Then I suggest the following modified version of the rule:
> 
> "Formalizing a result or a group of results one should use the
> existing definitions wherever they are available. For example, if the
> ring of integers appears in the formulation of a theorem or in a
> definition of a more complex object the existing the definition of
> [hz] must be used.
> 
> If, for some reason, a different model of a construction is strongly
> preferable it must be proved to be equivalent to the existing one."


The point I was trying to make is that we don't need a rule like that -
neither the original nor your modified version -, and that it does more
harm than good.

More bluntly, I don't see how it does any good at all:
It will simply prevent people from filing pull requests, and instead
they will keep their alternative definitions to themselves. In this way,
UniMath does not benefit at all.

Your rule will not make people do what you want. It will only prevent
them from contributing back to UniMath.

**Comment by vladimirias** [comment]
I doubt that we should be overly concerned with having too few contributions to UniMath. I am more worried that in a few years we will have too many and that this mass of code will become amorphous and no one but the authors will know what is in their contributions and the authors of the new contributions will not know what has been done by previous authors.

We have this problem already - recall Dan’s second definition of the function stn n -> stn n+i .

BTW, such duplications as Dan’s one can be found automatically and it would be interesting to have such a tool an to run it over the existing code.

The rule that I suggest won’t eliminate duplications but at least it will give the authors an extra motivation to learn what has been done by previous authors.

**Comment by rmatthes** [comment]
I am sorry for this rather administrative remark, but it seems to me that the style guide in UniMath/UniMath/README.md is not enough exposed. In particular, in the README file on the toplevel, I read 'The "UniMath" subdirectory contains various packages of formalized
mathematics. Please see the file README (or README.md) in each package for more
information about its contents.'. This seems to suggest that only the packages contain the README information, while the toplevel of the UniMath *subdirectory* might be missed. This is eased by the same name for the parent directory.

**Comment by benediktahrens** [comment]
On 07/07/2016 01:15 PM, Ralph Matthes wrote:
> I am sorry for this rather administrative remark, but it seems to me
> that the style guide in UniMath/UniMath/README.md is not enough
> exposed. In particular, in the README file on the toplevel, I read
> 'The "UniMath" subdirectory contains various packages of formalized 
> mathematics. Please see the file README (or README.md) in each
> package for more information about its contents.'. This seems to
> suggest that only the packages contain the README information, while
> the toplevel of the UniMath *subdirectory* might be missed. This is
> eased by the same name for the parent directory.

You are right, the readme in UniMath/UniMath is easy to miss.

We should put a link to UniMath/UniMath/README.md into UniMath/README.md.

**Comment by benediktahrens** [comment]
I think that this issue is not really one. Can we close it?

**Comment by DanGrayson** [comment]
No, let's adopt VV's rule in https://github.com/UniMath/UniMath/issues/395#issuecomment-230313039

**Comment by DanGrayson** [comment]
Re: "If, for some reason, a different model of a construction is strongly preferable it must be proved to be equivalent to the existing one."

This seems to be a good requirement, because then any results formulated using the alternative model can be transported back to the standard model, making them usable by others.

**Comment by DanGrayson** [comment]
If we insist that alternative definitions be proved equivalent to existing ones, that will make a lot of work for us in `CategoryTheory/limits/graphs/`, where there are hundreds of such definitions.

**Comment by benediktahrens** [comment]
On 20/11/2018 11:41, Daniel R. Grayson wrote:
> If we insist that alternative definitions be proved equivalent to 
> existing ones, that will make a lot of work for us in 
> |CategoryTheory/limits/graphs/|, where there are hundreds of such 
> definitions.

Which pairs do you have in mind specifically?

**Comment by DanGrayson** [comment]
These:
```
ArrowsFromZero ArrowsToZero BinCoproductArrow BinCoproductArrowEta
BinCoproductArrowUnique BinCoproductIn1 BinCoproductIn1Commutes BinCoproductIn2
BinCoproductIn2Commutes BinCoproductObject BinCoproductOfArrows
BinCoproductOfArrowsIn1 BinCoproductOfArrowsIn2 BinCoproduct_endo_is_identity
BinCoproducts BinCoproducts_from_Colims BinProductArrow BinProductArrowEta
BinProductArrowUnique BinProductObject BinProductOfArrows BinProductOfArrowsPr1
BinProductOfArrowsPr2 BinProductPr1 BinProductPr1Commutes BinProductPr2
BinProductPr2Commutes BinProducts BinProducts_from_Lims Cocone_by_postcompose
Coequalizer CoequalizerArrow CoequalizerEndo_is_identity CoequalizerObject
CoequalizerOut Coequalizers Cokernel Colims ColimsFunctorCategory_of_shape
Colims_of_shape Cone_precompose Equalizer EqualizerArrow
EqualizerEndo_is_identity EqualizerIn EqualizerObject Equalizers Initial
InitialArrow InitialArrowUnique InitialEndo_is_identity InitialObject Kernel
LimCone LimFunctor LimFunctorCone LimFunctor_data LimFunctor_mor LimFunctor_ob
LimFunctor_unique Lims LimsFunctorCategory LimsFunctorCategory_of_shape
Lims_of_shape Pullback PullbackArrow PullbackArrowUnique PullbackArrowUnique'
PullbackArrow_PullbackPr1 PullbackArrow_PullbackPr2 PullbackEndo_is_identity
PullbackObject PullbackPr1 PullbackPr2 PullbackSqrCommutes Pullbacks Pushout
PushoutArrow PushoutArrowUnique PushoutArrow_PushoutIn1 PushoutArrow_PushoutIn2
PushoutEndo_is_identity PushoutIn1 PushoutIn2 PushoutObject PushoutSqrCommutes
Pushouts Terminal TerminalArrow TerminalArrowUnique TerminalEndo_is_identity
TerminalObject Zero ZeroArrow ZeroArrowFrom ZeroArrowTo ZeroEndo_is_identity
ZeroObject are_inverses_from_Coequalizer_to_Coequalizer
are_inverses_from_Equalizer_to_Equalizer are_inverses_from_Pullback_to_Pullback
are_inverses_from_Pushout_to_Pushout bincoproduct_diagram binproduct_diagram
cocone cone coneOut coneOutCommutes cone_pointwise diagram edge
from_BinCoproduct_to_BinCoproduct from_Coequalizer_to_Coequalizer
from_Equalizer_to_Equalizer from_Pullback_to_Pullback from_Pushout_to_Pushout
glueSquares graph hasBinCoproducts hasBinProducts hasCoequalizers hasEqualizers
hasInitial hasLims hasPullbacks hasPushouts hasTerminal hasZero
identity_is_Pullback_input identity_is_Pushout_input
inv_from_iso_iso_from_Pullback inv_from_iso_iso_from_Pushout isCoequalizer
isCoequalizer_Coequalizer isColimCocone isColim_weq isColim_weq_subproof1
isColim_weq_subproof2 isEqualizer isEqualizer_Equalizer isInitial isLimCone
isLimCone_LimCone isPullback isPullbackGluedSquare isPullback_Pullback
isPushout isPushout_Pushout isTerminal isZero is_functor_LimFunctor_data
is_iso_from_BinCoproduct_to_BinCoproduct isaprop_Colims isaprop_Lims
isaprop_Pullbacks isaprop_isLimCone isaprop_isPullback
isiso_from_Coequalizer_to_Coequalizer isiso_from_Equalizer_to_Equalizer
isiso_from_Initial_to_Initial isiso_from_Pullback_to_Pullback
isiso_from_Pushout_to_Pushout isiso_from_Terminal_to_Terminal
isiso_from_Zero_to_Zero iso_Initials iso_Terminals iso_Zeros
iso_from_BinCoproduct_to_BinCoproduct iso_from_Coequalizer_to_Coequalizer
iso_from_Equalizer_to_Equalizer iso_from_Pullback_to_Pullback
iso_from_Pushout_to_Pushout iso_from_lim_to_lim lim limArrow limArrowCommutes
limArrowEta limArrowUnique limCone limOfArrows limOfArrowsOut limOut
limOutCommutes limUnivProp lim_endo_is_identity lim_nat_trans_in_data
mk_Coequalizer mk_Equalizer mk_Initial mk_LimCone mk_Pullback mk_Pushout
mk_Terminal mk_Zero mk_cone mk_isCoequalizer mk_isEqualizer mk_isInitial
mk_isPullback mk_isPushout mk_isTerminal mk_isZero
postCompWithLimOfArrows_subproof postcompWithBinCoproductArrow
postcompWithBinProductArrow precompWithBinCoproductArrow
precompWithBinProductArrow preserves_colimit transportf_isotoid' two_graph
uniqueExists
```
